src/pvcam.cpp

```diff
@@ -0,0 +1,1279 @@
+const std::unordered_map<SampleType, PL_IMAGE_FORMATS>
+  sample_type_to_pixel_format{
+      { SampleType_u8, PL_IMAGE_FORMAT_MONO8 },
+      { SampleType_u16, PL_IMAGE_FORMAT_MONO16 },
+  };
+
+const std::unordered_map<PL_IMAGE_FORMATS, SampleType>
+  pixel_format_to_sample_type{
+      { PL_IMAGE_FORMAT_MONO8, SampleType_u8 },
+      { PL_IMAGE_FORMAT_MONO16, SampleType_u16 },
+  };
+
+/**
+ * @brief Checks that the given camera id is within the range of a signed
+ * 16-bit.
+ * @param id The camera id to check.
+ * @throws std::runtime_error if the id is out of range.
+ */
+void
+check_pvcam_camera_id(uint64_t id)
+{
+    constexpr uint64_t limit = std::numeric_limits<int16>::max();
+    EXPECT(id <= limit, "Expected a value at most %llu. Got %llu.", limit, id);
+}
+
+void
+camera_properties_to_rgn_type(CameraProperties* properties, rgn_type& roi)
+{
+    auto binning = (uns16)properties->binning;
+    roi = {
+        .s1 = (uns16)properties->offset.x,
+        .s2 = (uns16)(properties->offset.x + properties->shape.x * binning - 1),
+        .sbin = binning,
+        .p1 = (uns16)properties->offset.y,
+        .p2 = (uns16)(properties->offset.y + properties->shape.y * binning - 1),
+        .pbin = binning,
+    };
+}
+
+void
+rgn_type_to_camera_properties(rgn_type& roi, CameraProperties* properties)
+{
+    properties->binning = (uint8_t)roi.pbin;
+
+    properties->offset = {
+        .x = roi.s1,
+        .y = roi.p1,
+    };
+    properties->shape = {
+        .x = (uint32_t)((roi.s2 - roi.s1 + 1) / properties->binning),
+        .y = (uint32_t)((roi.p2 - roi.p1 + 1) / properties->binning),
+    };
+}
+
+// forward
+struct PVCamCamera;
+
+struct CallbackContext
+{
+    int16 hcam;
+
+    std::shared_ptr<std::mutex> api_mutex;
+    std::shared_ptr<std::mutex> frame_mutex;
+    std::condition_variable cv;
+    std::atomic<int> frames_available;
+
+    uint8_t** frame;
+    FRAME_INFO info;
+};
+
+void
+callback_handler(FRAME_INFO* frame_info, void* context)
+{
+    auto* ctx = (CallbackContext*)context;
+    std::unique_lock<std::mutex> frame_lock(*ctx->frame_mutex);
+
+    PVCAM(pl_exp_get_oldest_frame_ex(ctx->hcam, (void**)ctx->frame, frame_info),
+          *ctx->api_mutex);
+
+    ctx->info = *frame_info;
+    ++ctx->frames_available;
+    ctx->cv.notify_one();
+}
+
+/// Camera declaration
+
+struct PVCamCamera final : public Camera
+{
+  public:
+    PVCamCamera(int16 hcam, std::shared_ptr<std::mutex> api_mutex);
+    ~PVCamCamera() noexcept;
+
+    void set(CameraProperties* properties);
+    void get(CameraProperties* properties);
+    void get_meta(CameraPropertyMetadata* meta) const;
+    void get_shape(ImageShape* shape) const;
+    void start();
+    void stop();
+    void execute_trigger() const;
+    void get_frame(void* im, size_t* nbytes, ImageInfo* info);
+
+  private:
+    // Guards access to the PVCAM API.
+    std::shared_ptr<std::mutex> pvcam_api_mutex_;
+
+    // The PVCAM camera handle.
+    int16 hcam;
```

Comment:
Why doesn't this private member have a trailing underscore?

```suggestion
    int16 hcam_;
```

Updates elsewhere will be needed, so you probably don't want to actually suggestion this suggestion through github.

Reply:
Thanks!